tests/framework/prototype.py

```diff
@@ -0,0 +1,208 @@
+#   [Step1] exec(): Execute a command to trigger the CREvent. For example, create a CR by a
+#                  `kubectl apply` command.
+#   [Step2] wait(): Wait for the system to converge.
+#   [Step3] checkRuleSets(): When the system converges, check all registered RuleSets.
+class CREvent:
+    def __init__(self, cr, cmd, ruleSets: list[RuleSet], timeout, namespace):
+        self.ruleSets = ruleSets
+        self.cmd = cmd
+        self.timeout = timeout
+        self.cr = cr
+        self.namespace = namespace
+    def trigger(self):
+        self.exec()
+        self.wait()
+        self.checkRuleSets()
+    def exec(self):
+        os.system(self.cmd)
+    def wait(self):
+        time.sleep(self.timeout)
+    def checkRuleSets(self):
+        for rs in self.ruleSets:
+            rs.checkRuleSet(self.cr, self.namespace)
+
+'''
+My implementations
+'''
+class SimpleMutator(Mutator):
+    def mutate(self):
+        for deltaSet in self.deltaSets:
+            steps = deltaSet.find_path(self.baseCR)
+            if steps:
+                for delta in deltaSet.iterate():
+                    yield self.apply_delta(self.baseCR, delta, steps)
+            else:
+                print("The path does not exist")
+
+class HeadPodNameRule(Rule):
+    def trigger_condition(self, cr) -> bool:
+        steps = "spec.headGroupSpec".split('.')
+        return (search_path(cr, steps) != None)
+
+    def assertRule(self, cr, namespace):
+        expected_val = search_path(cr, "spec.headGroupSpec.template.spec.containers.0.name".split('.'))
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace = namespace, label_selector='rayNodeType={}'.format('head'))
+        assert(headpods.items[0].spec.containers[0].name == expected_val)
+
+class EasyJobRule(Rule):
+    def assertRule(self, cr=None, namespace='default'):
+        headpods = client.CoreV1Api().list_namespaced_pod(namespace = namespace, label_selector='rayNodeType={}'.format('head'))
+        headpodName = headpods.items[0].metadata.name
+        rtn = os.system("kubectl exec {} -- python -c \"import ray; ray.init(); print(ray.cluster_resources())\"".format(headpodName))
+        assert(rtn == 0)
+
+if __name__ == '__main__':
+    template_name = 'config/ray-cluster.mini.yaml.template'
+    namespace = 'default'
+    with open(template_name) as base_yaml:
+        baseCR = yaml.load(base_yaml, Loader=yaml.FullLoader)
+    ds = DeltaSet("spec.headGroupSpec.template.spec.containers.0.name", ['ray-head-1', 'ray-head-2', 'ray-head-3'])
+    rs = RuleSet([HeadPodNameRule(), EasyJobRule()])
+    mut = SimpleMutator(baseCR, [ds])
+    images = ['rayproject/ray:2.0.0', 'kuberay/operator:v0.3.0', 'kuberay/apiserver:v0.3.0']
+
+    for cr in mut.mutate():
+        # Convert CR object into YAML file
+        filename = "tmp.yaml"
+        yaml_to_file(cr, filename)
+        # Prepare for KinD cluster, CRD, images, and Kuberay operator.
+        delete_kind_cluster()
```

Comment:
Setup and teardown operations you'd want to do in pytest fixtures.

Reply:
Updated https://github.com/ray-project/kuberay/pull/605/commits/893b38a2482ca53d340bb7e02f640263608ce0e3.